### PR TITLE
Explained the need to support database downgrading

### LINKED ISSUE #6315: Investigate the need to support database downgrading
#### Problem description
Our databases (forms, instances, savepoints, entities) currently support downgrading through the `#onDowngrade` method, which simply recreates the database from scratch. However, this feature may be unnecessary since Android does not allow app downgrades (you must uninstall the newer version before installing an older one).

We began considering this issue in [this discussion](https://github.com/getodk/collect/pull/6290#discussion_r1706683317), so it would be worthwhile to investigate whether we can eliminate the downgrading mechanism entirely or if it is still necessary (for example, if downgrading is possible using ADB or other methods).

### PR DESCRIPTION
**Pull request by grzesiek2010**
Closes #6315 

#### Why is this the best possible solution? Were any other approaches considered?
We have discussed the issue and agreed that since downgrading is not officially supported we should stop handling it in our codebase.

#### How does this change affect users? Describe intentional changes to behavior and behavior that could have accidentally been affected by code changes. In other words, what are the regression risks?
Downgrading isn't officially supported, so we don't typically test for it. I believe we can skip testing in this case as well. A code review by @seadowg.

#### Do we need any specific form for testing your changes? If so, please attach one.
No.

#### Does this change require updates to documentation? If so, please file an issue [here]( https://github.com/getodk/docs/issues/new) and include the link below.
No.

#### Before submitting this PR, please make sure you have:
- [x] added or modified tests for any new or changed behavior
- [x] run `./gradlew connectedAndroidTest` (or `./gradlew testLab`) and confirmed all checks still pass
- [x] added a comment above any new strings describing it for translators
- [x] added any new strings with date formatting to `DateFormatsTest`
- [x] verified that any code or assets from external sources are properly credited in comments and/or in the [about file](https://github.com/getodk/collect/blob/master/collect_app/src/main/assets/open_source_licenses.html).
- [x] verified that any new UI elements use theme colors. [UI Components Style guidelines](https://github.com/getodk/collect/blob/master/docs/CODE-GUIDELINES.md#ui-components-style-guidelines)

### CODE REVIEW

File: androidshared/src/main/java/org/odk/collect/androidshared/sqlite/DatabaseMigrator.java [hunk 1]
```diff
@@ -7,5 +7,11 @@ public interface DatabaseMigrator {
 
     void onUpgrade(SQLiteDatabase db, int oldVersion);
 
+    /**
+     * Downgrading an app is generally not officially supported without first uninstalling
+     * the current version. However, it may be possible in some cases, depending on the app
+     * and device. Rooted devices, for example, might allow more direct downgrades.
```
Comment (seadowg) [inline androidshared/src/main/java/org/odk/collect/androidshared/sqlite/DatabaseMigrator.java]: Thanks for looking into this! Seeing this, I'm actually leaning towards the oppositive and just exploding if we get a call to `onDowngrade`. I think the idea that data gets cleared out might be more dangerous than just crashing and letting the user export the data if they still need it or at least understand that they're in a bad state. Would be good to get input from @lognaturel on this as well.

**Review by seadowg — CHANGES_REQUESTED** [review]

File: androidshared/src/main/java/org/odk/collect/androidshared/sqlite/DatabaseMigrator.java (hunk 1 above)
Comment (grzesiek2010) [reply androidshared/src/main/java/org/odk/collect/androidshared/sqlite/DatabaseMigrator.java]: Interesting let's see what @lognaturel thinks.

Comment (lognaturel) [reply androidshared/src/main/java/org/odk/collect/androidshared/sqlite/DatabaseMigrator.java]: Here's an error message that I think would be helpful to someone in this state: "Downgrading ODK Collect is not supported. If you need the forms and data that ODK Collect currently manages, you must upgrade to the version you used previously or a newer version. If you don't need the data, you can uninstall this version and then reinstall it again."

Does that sound right? If so, let's go with something like that. It doesn't need to be translated.

**Review by seadowg — CHANGES_REQUESTED** [review]
As discussed on a call, we're going to just throw an exception with a message like "App downgrade not supported" if `onDowngrade` is called.

**Comment by grzesiek2010** [comment]
> As discussed on a call, we're going to just throw an exception with a message like "App downgrade not supported" if `onDowngrade` is called.

This exception is thrown by default by `SQLiteOpenHelper` if `onDowngrade` is not implemented:
```java
throw new SQLiteException("Can't downgrade database from version " +
                oldVersion + " to " + newVersion);
```
I think it's enough we don't need to change it.

**Review by seadowg — APPROVED** [review]

**Pull request merged**